Bio/SeqIO/AbiIO.py

```diff
@@ -33,23 +33,250 @@
 # each key is tag_name + tag_number
 # if a tag entry needs to be added, just add its key and its key
 # for the annotations dictionary as the value
+# dictionary for tags that require preprocessing before use in creating
+# seqrecords
 _EXTRACT = {
     'TUBE1': 'sample_well',
     'DySN1': 'dye',
     'GTyp1': 'polymer',
     'MODL1': 'machine_model',
 }
-# dictionary for tags that require preprocessing before use in creating
-# seqrecords
-_SPCTAGS = [
-    'PBAS2',    # base-called sequence
-    'PCON2',    # quality values of base-called sequence
-    'SMPL1',    # sample id inputted before sequencing run
-    'RUND1',    # run start date
-    'RUND2',    # run finish date
-    'RUNT1',    # run start time
-    'RUNT2',    # run finish time
-]
+
+
+# Complete data structure representing 98% of the API. The general section
+# represents the part of the API that's common to ALL instruments, whereas the
+# instrument specific sections are labelled as they are in the ABIF spec
+#
+# The 2%:
+#  - on some fields, there are repeated fields (e.g DYEN1-100), and those are not supported
+#  - OvrI1-N containing color data of some sort
+#
+# Keys don't seem to clash from machine to machine, so when we parse, we look
+# for ANY key, and store that in the raw ABIF data structure attached to the
+# annotations, with the assumption that anyone parsing the data can look up the
+# spec themself
+#
+# Key definitions are retained in case end users want "nice" labels pre-made
+# for them for all of the available fields.
+_INSTRUMENT_SPECIFIC_TAGS = {
+    'general': {
+        'APFN2': 'Sequencing Analysis parameters file name',
+        'APXV1': 'Analysis Protocol XML schema version',
+        'APrN1': 'Analysis Protocol settings name',
+        'APrV1': 'Analysis Protocol settings version',
+        'APrX1': 'Analysis Protocol XML string',
+        'CMNT1': 'Sample Comment',
+        'CTID1': 'Container Identifier, a.k.a. plate barcode',
+        'CTNM1': 'Container name, usually identical to CTID, but not necessarily so',
+        'CTTL1': 'Comment Title',
+        'CpEP1': 'Capillary type electrophoresis. 1 for a capillary based machine. 0 for a slab gel based machine.',
+        'DATA1': 'Channel 1 raw data',
+        'DATA1': 'Channel 1 raw data',
```

Comment:
I agree with Bow - they look like duplicates. On another note, we could
have additional useful tags available if we have the program check which
ABI version is being used and then add the tags specific for that version.
This would not take up much running time and would allow for tags like POS,
which are required to generate plots of the sequence and evaluate
heterozygous mutations. Let me know if you need more test files as I have
hundreds of files with various mutations and quality scores available.
Unfortunately, I do not have test files from all the different ABI
versions, which would be useful if a couple lines of code are added to
check the version and read the file accordingly.

All the Best,
David

On Thu, Feb 12, 2015 at 7:39 AM, Wibowo Arindrarto <notifications@github.com
> wrote:

> In Bio/SeqIO/AbiIO.py
> <https://github.com/biopython/biopython/pull/473#discussion_r24577415>:
>
> > +# Key definitions are retained in case end users want "nice" labels pre-made
> > +# for them for all of the available fields.
> > +_INSTRUMENT_SPECIFIC_TAGS = {
> > +    'general': {
> > +        'APFN2': 'Sequencing Analysis parameters file name',
> > +        'APXV1': 'Analysis Protocol XML schema version',
> > +        'APrN1': 'Analysis Protocol settings name',
> > +        'APrV1': 'Analysis Protocol settings version',
> > +        'APrX1': 'Analysis Protocol XML string',
> > +        'CMNT1': 'Sample Comment',
> > +        'CTID1': 'Container Identifier, a.k.a. plate barcode',
> > +        'CTNM1': 'Container name, usually identical to CTID, but not necessarily so',
> > +        'CTTL1': 'Comment Title',
> > +        'CpEP1': 'Capillary type electrophoresis. 1 for a capillary based machine. 0 for a slab gel based machine.',
> > +        'DATA1': 'Channel 1 raw data',
> > +        'DATA1': 'Channel 1 raw data',
>
> These look like duplicate key-value pairs ;)...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biopython/biopython/pull/473/files#r24577415>.
>